tsl/src/compression/api.c

```diff
@@ -892,3 +896,584 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+			if (!segment_info_datum_is_in_group(current_segment[seg_idx++]->segment_info,
+												val,
+												is_null))
+			{
+				/* new segment, need to do per-segment processing */
+				pfree(current_segment[seg_idx - 1]->segment_info); // bacause increased previously
+				SegmentInfo *segment_info =
+					segment_info_new(TupleDescAttr(slot->tts_tupleDescriptor, i));
+				segment_info_update(segment_info, val, is_null);
+				current_segment[seg_idx - 1]->segment_info = segment_info;
+				current_segment[seg_idx - 1]->decompressed_chunk_offset =
+					per_cols[i].decompressed_column_offset;
+			}
+		}
+	}
+}
+
+static bool
+decompress_segment_changed_group(CompressedSegmentInfo **current_segment, TupleTableSlot *slot,
+								 PerCompressedColumn **per_cols)
+{
+	PerCompressedColumn *decompressor_per_compressed_cols = *per_cols;
+	Datum val;
+	bool is_null;
+	bool changed_segment = false;
+	int seg_idx = 0;
+	for (int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
+	{
+		if (!compressed_chunk_column_is_segmentby(decompressor_per_compressed_cols[i]))
+			continue;
+		else
+		{
+			val = slot_getattr(slot, AttrOffsetGetAttrNumber(i), &is_null);
+			if (!segment_info_datum_is_in_group(current_segment[seg_idx++]->segment_info,
+												val,
+												is_null))
+			{
+				changed_segment = true;
+				break;
+			}
+		}
+	}
+	return changed_segment;
+}
+
+static bool
+recompress_remote_chunk(FunctionCallInfo fcinfo, Chunk *chunk)
+{
+	return invoke_compression_func_remotely(fcinfo, chunk);
+}
+
+/*
+ * This is hacky but it doesn't matter. We just want to check for the existence of such an index
+ * on the compressed chunk. For distributed hypertables, returning the index oid would raise issues,
+ * because the Access Node does not see that oid. So we return the oid of the uncompresed chunk
+ * instead, when an index is found.
+ */
+extern Datum
+tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (NULL == uncompressed_chunk)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
+
+	/* push down to data nodes for distributed case */
+	if (uncompressed_chunk->relkind == RELKIND_FOREIGN_TABLE)
+	{
+		/* look for index on data nodes */
+		if (!(invoke_compression_func_remotely(fcinfo, uncompressed_chunk)))
+			PG_RETURN_NULL();
+		else // don't care what the idx oid is, data node will find it and open it
+			PG_RETURN_OID(uncompressed_chunk_id);
+	}
+	int32 srcht_id =
+		uncompressed_chunk->fd.hypertable_id; // need it to find the segby cols from the catalog
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
+
+	List *htcols_list = ts_hypertable_compression_get(srcht_id);
+	ListCell *lc;
+	int htcols_listlen = list_length(htcols_list);
+	// find segmentby columns (need to know their index)
```

Comment:
I'm not sure what the outcome of the discussion about the comments was, but I would suggest using consistent comment characters within a function.

```suggestion
	/* find segmentby columns (need to know their index) */
```

Reply:
The outcome was that I will look for formatting tools we can use and then create a poll for us to decide, but I haven't done this yet. Regarding the comments in this PR, I will fix them all to follow the `/**/` style